core/src/layers/blocking.rs

```diff
@@ -0,0 +1,262 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use bytes;
+use bytes::Bytes;
+use tokio::runtime;
+
+use crate::raw::oio::ReadExt;
+use crate::raw::*;
+use crate::*;
+
+/// Add blocking API support for every operations.
+///
+/// # Blocking API
+///
+/// - This layer is auto-added to the operator if it's accessor doesn't support blocking APIs.
+///
+/// Tracking issue: #2678
+#[derive(Debug, Clone)]
+pub struct BlockingLayer {
+    runtime: Arc<runtime::Runtime>,
+}
+
+impl Default for BlockingLayer {
+    fn default() -> Self {
+        let runtime = runtime::Builder::new_current_thread()
```

Comment:
Or maybe we could use the Handle to the current runtime? Tried this and it works

Reply:
> Or maybe we could use the Handle to the current runtime?

It's possible that we don't have an async runtime. Maybe we need to check the result of [`try_current`](https://docs.rs/tokio/latest/tokio/runtime/struct.Handle.html#method.try_current)?

Reply:
> > Or maybe we could use the Handle to the current runtime?
> 
> It's possible that we don't have an async runtime. Maybe we need to check the result of [`try_current`](https://docs.rs/tokio/latest/tokio/runtime/struct.Handle.html#method.try_current)?

As the `default()` can't return a `Result`, which means the user can't handle this error, I prefer to not implement the `Default` trait for `BlockingLayer`, just provide a `with_handle(Handle) -> Self` method

Reply:
> I prefer to not implement the `Default` trait for `BlockingLayer`, just provide a `with_handle(Handle) -> Self` method

Removing `default()` is ok, since it's not a requirement for `Layer`.

But adding a new `with_handle()` API for `Layer` since a bit tricky. We don't want to expose this as our public API.

Reply:
> > I prefer to not implement the `Default` trait for `BlockingLayer`, just provide a `with_handle(Handle) -> Self` method
> 
> Removing `default()` is ok, since it's not a requirement for `Layer`.
> 
> But adding a new `with_handle()` API for `Layer` since a bit tricky. We don't want to expose this as our public API.

I added it as `BlockingLayer`'s method, but not `Layer`, just like the `LoggingLayer`'s `with_error_level`, is this OK?

Reply:
BlockingLayer's method is also OpenDAL's public API. How about do those thing at `BlockingLayer::create() -> Result<Self>`?